src/persist-client/src/impl/state.rs

```diff
@@ -568,4 +578,44 @@ mod tests {
             .collect::<Vec<_>>();
         assert_eq!(actual, vec!["key1", "key2", "key3"]);
     }
+
+    #[test]
+    fn compare_and_append() {
+        mz_ore::test::init_logging();
+        let mut state = State::<String, String, u64, i64>::new(ShardId::new()).collections;
+
+        // State is initally empty.
+        assert_eq!(state.trace.len(), 0);
+
+        // Cannot insert a batch with a lower != current shard upper.
+        assert_eq!(
+            state.compare_and_append(&["key1".to_owned()], &desc(1, 2)),
+            Break(Ok(Upper(Antichain::from_elem(0))))
+        );
+
+        // Insert an empty batch over a nonempty interval to move the shard upper
+        // to a nonzero value.
+        assert_eq!(state.compare_and_append(&[], &desc(0, 5)), Continue(()));
+
+        // Cannot insert a batch with a upper less than the lower.
+        assert_eq!(
+            state.compare_and_append(&["key1".to_owned()], &desc(5, 4)),
+            Break(Err(InvalidBounds {
+                lower: Antichain::from_elem(5),
+                upper: Antichain::from_elem(4)
+            }))
+        );
+
+        // Cannot insert a nonempty batch with an upper equal to lower.
```

Comment:
nit: use the same language here and the next assert to show they're related, so either "upper equal to lower" for both or "empty time interval"